NCrystal becomes runtime rather than buildtime dependency

tkittel opened the pull request:
<!--
If you are a first-time contributor to OpenMC, please have a look at our
contributing guidelines:
https://github.com/openmc-dev/openmc/blob/develop/CONTRIBUTING.md
-->

# Description

As discussed on #3304, this PR makes NCrystal a run-time rather than a build-time dependency, also for the C++ layer of openmc (it was always a run-time only dependency in the Python layer).

This has a lot of advantages, primarily:

- No hassle trying to matrix in NCrystal=yes and NCrystal=no in binary distributions like at conda and the incoming python wheels by @ahnaf-tahmid-chowdhury, and the ensuing headache of wondering about ABI stability and when a new NCrystal release will require a rebuild of OpenMC packages.
- People using NCrystal will no longer be forced to perform a custom rebuild of OpenMC, they simply have to also install the ncrystal package. So any build of OpenMC (conda, python wheels, manual builds) will work with any build of NCrystal (conda, python wheels, manual builds), assuming just a simple platform compatibility (i.e. you still can't be silly and try to mix openmc-aarch64 packages with ncrystal-x86 of course).

And note that for users not adding NCrystal materials to their OpenMC geometries, there should be ~0 overhead. And people adding NCrystal materials to their geometries will get a nice pedagogical error message if NCrystal is not already installed, telling them to install it. Once they have done so, they can simply rerun and it should just work.

So on the NCrystal / ESS side we will be super happy with this, since it means we can use OpenMC+NCrystal out of the box, and hopefully other OpenMC developments will also become easier going forward, since there will be less of a need to worry about configuration and build issues related to NCrystal.

The technology behind all of this is simply to load NCrystal as a dynamic plugin with `dlopen`, like I can see is already done for compiled sources in OpenMC. Thus, the `-ldl` (from `CMAKE_DL_LIBS`) is already used for OpenMC builds, and there is no new dependency introduced there.

The ability to actually load and use NCrystal via an abstract plugin interface was introduced in the recent [NCrystal 4.1.0 release](https://github.com/mctools/ncrystal/releases/tag/v4.1.0), and has been already tested to work in NCrystal's CI on macOS/Ubuntu/Windows/aarch64/intel, so I am pretty confident that this is solid :-)

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
<!--
While tests will automatically be checked by CI, it is good practice to
ensure that they pass locally first. See instructions here:
https://docs.openmc.org/en/latest/devguide/tests.html
-->

tkittel commented:
@shimwell @ahnaf-tahmid-chowdhury @paulromano FYI.

Note that I converted the PR to draft for now, since there might be a few issues related to `docs/source/usersguide/install.rst` and doxygen comments. However, feel free to comment already :-)

Cheers,
Thomas

tkittel commented:
I forgot to mention, the infrastructure for discovering NCrystal (via the `ncrystal-config` command) and calling out to `dlopen`+`dlsym` or `LoadLibrary`+`GetProcAdress`, and wrapping the abstract NCrystal interface into a small helper class `NCrystalScatProc` is all encapsulated in two new files `ncrystal_load.h` and `ncrystal_load.cpp`.

marquezj commented:
Thanks @tkittel. This will simplify a lot the deployment of OpenMC with NCrystal materials.

From what I see you need to remove the CMake option for NCrystal in `tools/ci/gha-install.py`.

tkittel commented:
Thanks @marquezj .

I just wanted to verify that all CI passed with the current commits, which it actually did. I removed ncrystal lines in that file + I spotted a few that I missed in `src/output.cpp` (basically ncrystal is no longer a build option, so there is no need to accept or print flags related to whether or not ncrystal is enabled).

tkittel commented:
I still need to look into doxygen strings in the changed C++ code, but apart from that I feel this is ready for review (unless my last two commits broke the tests, which they should hopefully not).

tkittel commented:
So I have tried to add some doxygen comments now, hope all is OK :-)

tkittel commented:
Could someone let me know:

1. Is the CODEOWNERS file in a correct format now? (ncrystal_interface.cpp is no longer the only ncrystal file, there is also ncrystal_load.h/.cpp).
2. How can I actually view what doxygen does with the comments? I could not find any reference to the new code in the readthedocs output.

marquezj commented on CODEOWNERS:
```diff
@@ -60,7 +60,7 @@ Dockerfile @shimwell
 
 # NCrystal interface
-src/ncrystal_interface.cpp @marquezj
+src/ncrystal_interface.cpp @marquezj @tkittel
```
Add `src/ncrystal_load.cpp`

marquezj replied:
Also, @tkittel needs to get write access to the repo to be a maintainer.

marquezj reviewed (commented):
From my side the only remaining issue is the `CODEOWNERS` file.

It might be useful in the future to check that the interoperability also works in other platforms, but currently CI is only set up for Ubuntu.

tkittel commented:
Thanks @marquezj @paulromano for the feedback and invite. I have updated CODEOWNERS to also have `src/ncrystal_load.cpp`, and did a rebase for good measure. 

Concerning other platforms than ubuntu, I actually _have_ verified that the mechanism works on all platforms (Windows, macOS, Linux, Intel, Arm, ...), since before releasing NCrystal 4.1.0 with the new APIs, I added a standalone example [here](https://github.com/mctools/ncrystal/tree/main/examples/virtualapi_project), which essentially exercises the `ncrystal_load.h/.cpp` files (they are basically the same files as those submitted to OpenMC in this PR, apart from modifications for OpenMC coding style).

You can see a workflow exercising this at https://github.com/mctools/ncrystal/blob/main/.github/workflows/plugindb.yml (recent run [here](https://github.com/mctools/ncrystal/actions/runs/13542966440)), in which the small app is built without any build-time dependency on NCrystal, and then it is used with an NCrystal installation only at run-time. For fun I even build NCrystal wheels and the example app on separate machines, before running them together on a third machine.

So I fully expect that when/if OpenMC adds more CI platforms in the future, that this will all just work :-)

tkittel commented:
Ok, so now all tests passed again. Any more suggestions or comments, or is this ready to merge?

ahnaf-tahmid-chowdhury commented on tests/regression_tests/ncrystal/test.py:
```diff
@@ -6,12 +6,13 @@
 pytestmark = pytest.mark.skipif(
-    not openmc.lib._ncrystal_enabled(),
-    reason="NCrystal materials are not enabled.")
+    not shutil.which('ncrystal-config'),
```
Just a small suggestion, but not necessary. I think we can try directly importing `NCrystal` if possible. That may eliminate the need for `shutil`.

tkittel replied:
Thanks for reviewing and the suggestion @ahnaf-tahmid-chowdhury .

I don't know... you are right of course that trying an `import NCrystal` would also work, but somehow I feel it would add some complications (e.g. things happening in the NCrystal python initialisation would happen at this early stage rather than later when used by openmc.

So I think I will keep it as it is, if that's ok.

ahnaf-tahmid-chowdhury reviewed (commented):
Thank you very much for the update! It is truly very streamlined.

marquezj reviewed (approved):
Good work. I think it is ready to merge.

tkittel commented:
@paulromano after several years and iterations of considering this, I honestly do not think there is any better alternative than querying `ncrystal-config` for the library path. Any other solution would inevitably be specific to each deployment mechanism (e.g. potentially different for Python wheels, conda builds, manual builds of NCrystal, debian packages, freebsd ports, etc.). Or it would require users to mess around with fiddly bits like environment variables, adding stuff to xml, or flags to openmc or whatnot. On the NCrystal side, we are devoted to always ensuring that any installation of NCrystal will have a correct functioning `ncrystal-config`, since that is how we now have chosen we want to support it (and it is how the NCrystal python modules look for the shared library).

Is your concern that this would break non-NCrystal users' usage of OpenMC, or that it would not work for people using NCrystal in OpenMC? For the former, they would not trigger the relevant code-paths at all, so it should be a non-issue (and popen is posix + in windows std libs). For the latter, I would say let us go ahead and let us see if any unexpected problems would surface down the line. If it would become necessary for some very specific use-case to provide the information via a compile time flag to OpenMC or perhaps an environment variable, we could always add it then. However, I would prefer not to add more complicated options until there is more than a theoretical need for them (after all, there would have to be some CI tests of these options or we could not rely on them anyway... and we would have to maintain docs as well, etc.).

As I mentioned, the mechanism has been tested elsewhere on a lot of platforms and seems to work without issue so far.

Cheers,
Thomas

paulromano commented:
Thanks for the quick response and explanation @tkittel. My reluctance is more about the philosophical idea of OpenMC spawning some other executable, and of using non-standard headers in the C++ code, although I fully understand that popen, dlopen, etc. are practically guaranteed to be available on any conceivable system. And of course we already use dlopen for custom source routines. So, I may just have to accept that practicality beats purity here :smile: 

If it's OK with you, I might do a little refactoring on this branch to separate out the dynamic linking calls to be reusable for this and custom source routines (which would have the benefit of enabling them to be used on windows too).

tkittel commented:
@paulromano I fully understand your uneasiness, but I hope I can convince you nonetheless since I believe this will be the most head-ache free and convenient path forward :-)

And of course, feel free to further refactor on this branch, I will keep my hands off it in the meantime :-)

tkittel commented on src/ncrystal_load.cpp:
```diff
@@ -0,0 +1,152 @@
+    auto handle = LoadLibrary(cfg.shlibpath.c_str());
+#else
+    dlerror(); // clear previous errors
+    void* handle = dlopen(cfg.shlibpath.c_str(), RTLD_LOCAL | RTLD_LAZY);
```
That is on purpose, since my experience with dlclose is that it introduces a lot more problems than it solves. We are using dlopen/dlclose for plugins in NCrystal since some time now, and IIRC I had to abandon my initial efforts at cleaning up properly with dlclose since it introduced all sorts of hard-to-debug and hard to avoid issues related to symbols still being needed after dlclose was called. And for very little benefit (nicer valgrind output comes to mind?). I did run valgrind on my little standalone example using my original `ncrystal_load.h/cpp` and verified that all the various dynamically resources were otherwise offloaded as they should. Only some weird stuff inside libdl was left.

I also do not think that you can necessarily always expect to dlopen-dlclose-dlopen-dlclose the same library, at least I recall some cases where that proved troublesome.

paulromano replied:
Interesting. Yeah, I don't think we'd want repeated dlopen/dlclose. I was thinking more at the point that `openmc_finalize` is called. We do call `dlclose` for the custom source routines and it seems to work fine in that use case.

tkittel replied:
I am still a bit afraid to do it... I recall some nasty stuff related to loading plugins into NCrystal, like internal caches in NCrystal that were cleaned up at the end of the programme, and at that time the library with the symbols had already been offloaded. I don't recall the details, I only recall that I came to the conclusion that the super pure valgrind report (that I normally crave) was not worth all the problems arising from the dlclose call. I am suspecting that perhaps your source plugins are a bit simpler, or has less static storage or whatever, than NCrystal, and that might explain why you had no issues. Or perhaps I am doing something wrong somewhere that you guys are not doing wrong, but in any case the outcome is the same :-)

I mean, of course we can try, but even after a few trial runs showing no issues I would not really be confident that some weird issues would not crop up later. And it doesn't even actually guarantee to help (see for instance https://stackoverflow.com/a/27030236 - dlclose is allowed to essentially be ignored), so even then the gain is a bit... undefined.

It is ultimately your call of course, I am just a bit afraid I will later be called on to debug some weird issues arising due to this (and people will talk about at the water cooler about how that weird stuff that Thomas contributed always makes problems, or NCrystal is something people should be afraid to use ;-) ).

tkittel replied:
I looked a bit at your compiled source and usage of dlclose/dlopen. It could be that the reason you are using this with no issue is that you clean up all the allocations from your source before calling dlclose. For instance, I see:

```c++
CompiledSourceWrapper::~CompiledSourceWrapper()
{
  // Make sure custom source is cleared before closing shared library
  if (compiled_source_.get())
    compiled_source_.reset();

#ifdef HAS_DYNAMIC_LINKING
  dlclose(shared_library_);
#else
  fatal_error("Custom source libraries have not yet been implemented for "
              "non-POSIX systems");
#endif
}
```
Which makes sense. However, libNCrystal.so (by whatever name) keeps a lot of allocated objects in static caches internally, so it would in general be hard to be sure to have everything flushed correctly (it gets worse if NCrystal was itself loading a custom plugin which itself might have some caches).

tkittel commented on src/ncrystal_load.cpp:
```diff
@@ -0,0 +1,152 @@
+    symbol += "_access_virtual_api";
+
+#ifdef NCLOAD_WINDOWS
+    FARPROC fproc;
```
Looks like it. I wonder why my compilation at windows did not emit any warnings. I will remove it.

tkittel commented on src/ncrystal_load.cpp:
```diff
@@ -0,0 +1,152 @@
+#include <stdio.h>
+#include <string>
+
+#if !defined(NCLOAD_WINDOWS) && (defined(_WIN32) || defined(WIN32))
```
Yeah, I recall spent an hour reading long discussions at stackoverflow and convinced myself that this was safer in general :-)

However, I am by no means 100% sure about this, so if you think _WIN32 only is enough or better, I am happy to change it?

paulromano replied:
Haha, I was also trying to read up to understand things and as best as I can tell `_WIN32` should cover things well enough. Why don't we adopt that and if our millions of Windows users complain we can always update it later :smile:

tkittel replied:
That's a sound policy ;-)

Ok, I'll change it as well. I am holding off a bit to bundle these tiny changes, since the CI takes 3 hours and every little push restarts it.

tkittel replied:
Btw., grepping for WIN32 I also see `__WIN32__` in src/external/Faddeeva.cc ;-)

tkittel replied:
(it is essentially very confusing... and we are not even talking about `__CYGWIN` yet ;-) )

paulromano reviewed (approved):
Alright, I think I'm satisfied with where this is. Thanks for working through my comments @tkittel. Can you take a quick look through my last changes in ncrystal_load.cpp simplifying the ifdefs and make sure you're OK with it?

tkittel commented:
Thanks @paulromano ! You changes look very OK, you even found another unused variable.

Which is surprising since I thought I had actually tried to compile with `-pedantic -Wall -Wextra`, but apparently not. I did so now and found *another* unused variable left over. So I fixed that as well, and now there really should be no more! Geez.

Btw., can I ask for the comments you left next to each include concerning what they bring in, are you using some code inspection tool, or do you simply have all the header files and the C++ standard in your head? ;-)

Anyway, all looks OK for me!

paulromano commented:
> can I ask for the comments you left next to each include concerning what they bring in, are you using some code inspection tool, or do you simply have all the header files and the C++ standard in your head?

I do it manually, usually consulting online documentation like https://devdocs.io/cpp. I wouldn't be surprised if there are automated ways to do this though! I don't personally use it but maybe something like [include-what-you-use](https://github.com/include-what-you-use/include-what-you-use) can do this?
